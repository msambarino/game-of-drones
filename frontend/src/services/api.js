import axios from 'axios'

axios.defaults.baseURL = process.env.REACT_APP_API_SERVER

export const fetchHighScores = () => {
  return axios.get('/scores')
}

export const registerScore = (player1, player2, winner) => {
  return axios.post('/scores', { player1, player2, winner })
}
