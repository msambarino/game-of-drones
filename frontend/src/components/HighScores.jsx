import React from 'react'
import { fetchHighScores } from 'services/api'
import { Link } from 'react-router-dom'

class HighScores extends React.Component {
  constructor(props) {
    super(props)

    this.state = {
      loading: false,
      error: null,
      ranking: []
    }

    this.renderScores = this.renderScores.bind(this)
  }

  componentDidMount() {
    this.setState({ loading: true })

    fetchHighScores()
      .then(response =>
        this.setState({ loading: false, ranking: response.data.data })
      )
      .catch(error => this.setState({ loading: false, error: true }))
  }

  renderScores() {
    const ranking = this.state.ranking
    return ranking && ranking.length > 0 ? (
      <div className="columns is-centered">
        <div className="column is-half">
          <table className="table is-fullwidth is-striped">
            <thead>
              <tr>
                <th />
                <th>Player</th>
                <th>Games won</th>
              </tr>
            </thead>
            <tbody>
              {ranking.map((rank, index) => (
                <tr key={index}>
                  <td>{index + 1}</td>
                  <td>{rank.player}</td>
                  <td>{rank.count}</td>
                </tr>
              ))}
            </tbody>
          </table>
        </div>
      </div>
    ) : (
      <div>
        <span>We didn't find any scores, </span>
        <Link to="/">Go play some games!</Link>
      </div>
    )
  }

  render() {
    const { loading, error } = this.state
    return (
      <div>
        {error ? (
          <div className="notification is-danger">
            Unable to fetch high scores
          </div>
        ) : null}
        <div className="container">
          <section className="section has-text-centered">
            <h1 className="title is-loading">High Scores</h1>
            {loading ? (
              <div className="loader" style={{ margin: 'auto' }} />
            ) : error ? null : (
              this.renderScores()
            )}
          </section>
        </div>
      </div>
    )
  }
}

export default HighScores
