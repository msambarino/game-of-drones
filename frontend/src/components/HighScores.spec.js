import React from 'react'
import { shallow, mount, render } from 'enzyme'
import { shallowToJson } from 'enzyme-to-json'
import HighScores from './HighScores'
import { fetchHighScores } from 'services/api'

jest.mock('services/api', () => ({
  fetchHighScores: jest.fn(() =>
    Promise.resolve({
      data: {
        data: [
          {
            player: 'p1',
            count: 10
          },
          {
            player: 'p2',
            count: 3
          },
          {
            player: 'p3',
            count: 2
          }
        ]
      }
    })
  )
}))

test('snapshot test', () => {
  const component = shallow(<HighScores />)
  const tree = shallowToJson(component)
  expect(tree).toMatchSnapshot()
})

test('shoud fetch high scores when mounted', () => {
  const component = shallow(<HighScores />)
  expect(fetchHighScores).toHaveBeenCalled()
})

test('should render a row for each high score', async () => {
  const component = shallow(<HighScores />)
  await component.instance().componentDidMount()
  component.update()
  expect(component.instance().state.ranking).toHaveLength(3)
  expect(component.find('tbody > tr')).toHaveLength(3)
})
