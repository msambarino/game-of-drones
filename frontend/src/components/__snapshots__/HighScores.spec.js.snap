// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`snapshot test 1`] = `
<div>
  <div
    className="container"
  >
    <section
      className="section has-text-centered"
    >
      <h1
        className="title is-loading"
      >
        High Scores
      </h1>
      <div
        className="loader"
        style={
          Object {
            "margin": "auto",
          }
        }
      />
    </section>
  </div>
</div>
`;
